SOFTWARE-3842 Installing a CMS XCache from RPM

efajardo opened the pull request:
This is part of [SOFTWARE-3842](https://opensciencegrid.atlassian.net/browse/SOFTWARE-3842) it is a place to start and I would like some comments from @piperov. 

@brianhlin I am waiting to add a special part if to use  redirector but I would like a review of [PR-75](https://github.com/opensciencegrid/xcache/pull/75).

piperov commented:
Looks like a good start

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+===============================
+
+This document describes how to install a CMS XCache.  This service allows a site or regional
+network to cache data frequently used in CMS, reducing data transfer over the wide-area network and
```
```suggestion
network to cache data frequently used by the [CMS experiment](https://github.com/opensciencegrid/topology/blob/master/virtual-organizations/CMS.yaml), reducing data transfer over the wide-area network and
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+* __Operating system:__ A RHEL 7 or compatible operating systems.
+* __User IDs:__ If they do not exist already, the installation will create the Linux user IDs `xrootd`
+* __Host certificate:__ Required for client authentication and authentication with CMS VOMS Server
+  Authenticated StashCache is an optional feature.
```
```suggestion
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+!!! note
+    The cache directory must be writable by the `xrootd:xrootd` user and group.
+
+The `stash-cache` package provides default configuration files in `/etc/xrootd/xrootd-stash-cache.cfg` and `/etc/xrootd/config.d/`.
```
```suggestion
The `cms-xcache` package provides default configuration files in `/etc/xrootd/xrootd-cms-xcache.cfg` and `/etc/xrootd/config.d/`.
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+- `set resourcename = YOUR_RESOURCE_NAME`: the resource name registered with the OSG for example ("T2_US_UCSD")
+
+!!! note
+    Xrootd can manage a set of independent disk for the cache. So you can modify file `90-cms-xcache-disks.cfg` and add the disks there then `/mnt/stash` just becomes a place to hold symlinks.
```
```suggestion
    XRootD can manage a set of independent disk for the cache. So you can modify file `90-cms-xcache-disks.cfg` and add the disks there then `rootdir` just becomes a place to hold symlinks.
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+
+### Ensure the xrootd service has a certificate
+
+The service will need a certificate for reporting and to authenticate to StashCache origins.
```
```suggestion
The service will need a certificate for reporting and to authenticate to CMS AAA.
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+    You must repeat the above steps whenever you renew your host certificate.
+    If you automate certificate renewal, you should automate copying as well.
+    For example, if you are using Certbot for Let's Encrypt, you should write a "deploy hook" as documented
+    [on the Certbot site](https://certbot.eff.org/docs/using.html#renewing-certificates).
```
```suggestion
    [here](/security/host-certs/lets-encrypt/#pre-and-post-renewal-hooks)
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+!!! note
+    You must repeat the above steps whenever you renew your host certificate.
+    If you automate certificate renewal, you should automate copying as well.
+    For example, if you are using Certbot for Let's Encrypt, you should write a "deploy hook" as documented
```
```suggestion
    For example, if you are using `certbot` for Let's Encrypt, you should write a "deploy hook" as documented
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+As soon as your issue has been resolved, revert any changes you have made to `/etc/xrootd/digauth.cfg`.
+
+
+Managing StashCache and associated services
```
```suggestion
Managing CMS XCache and associated services
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+| Enable a service to start on boot       | `systemctl enable <SERVICE-NAME>`  |
+| Disable a service from starting on boot | `systemctl disable <SERVICE-NAME>` |
+
+### CMS Xcache services
```
```suggestion
### CMS XCache services
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+
+### CMS Xcache cmsd (optional)
+
+In addition to the cache services if your cache is composed of several hosts then you need to run the service`cmsd`
```
```suggestion
In addition to the cache services if your cache is composed of several hosts then you need to run the service `cmsd`
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,182 @@
+Validating the Cache
+---------------------
+
+The cache server functions as a normal CMS Xrootd server so use first a personal CMS X509 voms proxy like this
```
```suggestion
The cache server functions as a normal CMS XRootD server so first verify it with a personal CMS X.509 proxy: 
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
````diff
@@ -0,0 +1,182 @@
+uri       : lcg-voms2.cern.ch:15002
+```
+
+Then test using xrdcp directly in your cache
````
```suggestion
Then test using `xrdcp` directly in your cache:
```

piperov commented:
I re-read the installation instructions, and they correspond to what we did here at Purdue, so I approve them. One possible clarification is that sites with multiple xcache servers can use a single SAN certificate for all servers.

efajardo commented:
@piperov I discussed this with @bbockelm this is only true for newer versions of CMSSW. Since we cannot guarantee this we will probably have to have one cert per node.

efajardo commented:
@brianhlin I implemented your suggestions and @piperov gave the good ahead. I also added the bits for horizontal scalling (redir) but that depends on this [PR](https://github.com/opensciencegrid/xcache/pull/75) getting merged and at least into osg development.

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,242 @@
+This document describes how to install a CMS XCache.  This service allows a site or regional
+network to cache data frequently used by the [CMS experiment](https://github.com/opensciencegrid/topology/blob/master/virtual-organizations/CMS.yaml), reducing data transfer over the wide-area network and
+decreasing access latency.
+
```
I'd still like a sentence here about the two different configurations: standalone + multi-node

efajardo replied:
ok fixed

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,242 @@
+
+## Creating a multinode cache (optional)
+
+Some sites would like to have a single logical cache composed of several nodes as shown bellow:
```
```suggestion
Some sites would like to have a single logical cache composed of several nodes as shown below:
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,242 @@
+| XRootD (Optional)| `cmsd@cms-xcache.service` | The xrootd daemon which performs authenticated data transfers |
+
+
+## Creating a multinode cache (optional)
```
```suggestion
Installing a Multinode Cache (optional)
---------------------------------------
```
Also, please move this above the managing services section

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,242 @@
+
+
+!!! note
+    If your are installing a multinode cache then instead of `yourlocalcache:1094` url should be changed for `yourcacheredirector:2040`
```
```suggestion
    If your are installing a [multinode cache](#installing-a-multinode-cache) then instead of `yourlocalcache:1094` url should be changed for `yourcacheredirector:2040`
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,242 @@
+| **Software** | **Service name** | **Notes** |
+|--------------|------------------|-----------|
+| XRootD (Optional)| `xrootd@xcache-redirector.service` | The xrootd daemon which performs authenticated data transfers |
+| XRootD (Optional)| `cmsd@cms-xcache.service` | The xrootd daemon which performs authenticated data transfers |
```
Is this right? I'd expect it to be `cmsd@xcache-redirector.service`

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,242 @@
+| **Software** | **Service name** | **Notes** |
+|--------------|------------------|-----------|
+| XRootD | `xrootd@cms-xcache.service` | The XRootD daemon, which performs the data transfers |
+| XRootD (Optional)| `cmsd@xcache-redirector.service` | The cmsd daemon that interact with the different xrootd servers |
```
Is this right? I'd expect it to be `cmsd@cms-xcache.service`

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,239 @@
+Remote debugging should only be enabled for as long as you need assistance.
+As soon as your issue has been resolved, revert any changes you have made to `/etc/xrootd/digauth.cfg`.
+
+## Installing a multinode cache (optional)
```
```suggestion
## Installing a Multinode Cache (optional)
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,239 @@
+
+This can be achieved by following the next steps
+
+### Setup an XCache redirector
```
```suggestion
### Install an XCache redirector
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,239 @@
+        :::file
+        all.manager yourlocalredir:2041
+
+1. Start the cmsd and xrootd proccess:
```
```suggestion
1. Start and enable the `cmsd` and `xrootd` services:
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,239 @@
+| XRootD | `cmsd@xcache-redirector.service` | The cmsd daemon that interact with the different xrootd servers |
+| XRootD | `xrootd@xcache-redirector.service` | The xrootd daemon which performs authenticated data transfers |
+
+1. Create a config file in the nodes where you installed your caches `/etc/xrootd/config.d/94-xrootd-manager.cfg` with the following contents:
```
```suggestion
### Configuring each of your cache nodes

1. Create a config file in the nodes where you installed your caches `/etc/xrootd/config.d/94-xrootd-manager.cfg` with the following contents:
```

brianhlin commented on docs/data/xrootd/install-cms-xcache.md:
```diff
@@ -0,0 +1,239 @@
+      :::file
+      all.manager yourlocalredir:2041
+
+1. Run the service `cmsd`
```
```suggestion
1. Start and enable the `cmsd` service:
```

brianhlin commented:
Overriding the link failure. Merging
